src/Common/Threading.cpp

```diff
@@ -0,0 +1,484 @@
+    {
+        return syscall(SYS_futex, address, FUTEX_WAIT_PRIVATE, value, nullptr, nullptr, 0);
+    }
+
+    inline Int64 futexWake(void * address, int count)
+    {
+        return syscall(SYS_futex, address, FUTEX_WAKE_PRIVATE, count, nullptr, nullptr, 0);
+    }
+
+    // inline void waitFetch(std::atomic<UInt32> & address, UInt32 & value)
+    // {
+    //     futexWait(&address, value);
+    //     value = address.load();
+    // }
+
+    // inline void wakeOne(std::atomic<UInt32> & address)
+    // {
+    //     futexWake(&address, 1);
+    // }
+
+    // inline void wakeAll(std::atomic<UInt32> & address)
+    // {
+    //      futexWake(&address, INT_MAX);
+    // }
+
+    inline constexpr UInt32 lowerValue(UInt64 value)
+    {
+        return static_cast<UInt32>(value & 0xffffffffull);
+    }
+
+    inline constexpr UInt32 upperValue(UInt64 value)
+    {
+        return static_cast<UInt32>(value >> 32ull);
+    }
+
+    inline UInt32 * lowerAddress(void * address)
+    {
+        return reinterpret_cast<UInt32 *>(address) + (std::endian::native == std::endian::big);
+    }
+
+    inline UInt32 * upperAddress(void * address)
+    {
+        return reinterpret_cast<UInt32 *>(address) + (std::endian::native == std::endian::little);
+    }
+
+    inline void waitLowerFetch(std::atomic<UInt64> & address, UInt64 & value)
+    {
+        futexWait(lowerAddress(&address), lowerValue(value));
+        value = address.load();
+    }
+
+    inline bool cancellableWaitLowerFetch(std::atomic<UInt64> & address, UInt64 & value)
+    {
+        bool res = CancelToken::local().wait(lowerAddress(&address), lowerValue(value));
+        value = address.load();
+        return res;
+    }
+
+    inline void wakeLowerOne(std::atomic<UInt64> & address)
+    {
+        syscall(SYS_futex, lowerAddress(&address), FUTEX_WAKE_PRIVATE, 1, nullptr, nullptr, 0);
+    }
+
+    // inline void wakeLowerAll(std::atomic<UInt64> & address)
+    // {
+    //     syscall(SYS_futex, lowerAddress(&address), FUTEX_WAKE_PRIVATE, INT_MAX, nullptr, nullptr, 0);
+    // }
+
+    inline void waitUpperFetch(std::atomic<UInt64> & address, UInt64 & value)
+    {
+        futexWait(upperAddress(&address), upperValue(value));
+        value = address.load();
+    }
+
+    inline bool cancellableWaitUpperFetch(std::atomic<UInt64> & address, UInt64 & value)
+    {
+        bool res = CancelToken::local().wait(upperAddress(&address), upperValue(value));
+        value = address.load();
+        return res;
+    }
+
+    // inline void wakeUpperOne(std::atomic<UInt64> & address)
+    // {
+    //     syscall(SYS_futex, upperAddress(&address), FUTEX_WAKE_PRIVATE, 1, nullptr, nullptr, 0);
+    // }
+
+    inline void wakeUpperAll(std::atomic<UInt64> & address)
+    {
+        syscall(SYS_futex, upperAddress(&address), FUTEX_WAKE_PRIVATE, INT_MAX, nullptr, nullptr, 0);
+    }
+}
+
+void CancelToken::Registry::insert(CancelToken * token)
+{
+    std::lock_guard<std::mutex> lock(mutex);
```

Comment:
`std::lock_guard lock(mutex)` (C++17 CTAD)

Reply:
not used much in our code, but I think we have to start changing it